### PR TITLE
support other prometheus formats

### LINKED ISSUE #3926: PrometheusExpositionService doesn't support OpenMetrics Format
The current exporter doesn't support OpenMetrics format. It's fairly simple and most of this is actually handled directly by the prometheus library itself via the TextFormat class. See: https://github.com/AngerM/ag_server/blob/main/base/src/main/kotlin/dev/angerm/ag_server/http/PrometheusHandler.kt 
Which I based off of https://github.com/prometheus/client_java/blob/master/simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java
to support some of the query options, etc.

### PR DESCRIPTION
**Pull request by AngerM**
Motivation:

Allows the prometheus exporter to write other formats

Modifications:

- Uses the TextFormat prometheus library to choose the format it writes in

Result:

- Closes #3926

### CODE REVIEW

**Review by minwoox — COMMENTED** [review]
Thanks a lot for the change. 😄 
Could you also add a test to `PrometheusExpositionServiceTest` so we can be sure that this change would work? 😆

**Comment by AngerM** [comment]
@minwoox updated

**Review by ikhoon — APPROVED** [review]
This is a quick fix. Thanks, @AngerM!

**Comment by AngerM** [comment]
@ikhoon The failed tests look unrelated to these changes. The linter errors I will fix

File: core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java [hunk 1]
```diff
@@ -85,11 +84,17 @@ public PrometheusExpositionService(CollectorRegistry collectorRegistry) {
...
+        final String accept = req.headers().get(HttpHeaderNames.ACCEPT);
+        final String format = TextFormat.chooseContentType(accept);
         final ByteArrayOutputStream stream = new ByteArrayOutputStream();
         try (OutputStreamWriter writer = new OutputStreamWriter(stream)) {
-            TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
+            TextFormat.writeFormat(
+                format,
+                writer,
+                collectorRegistry.metricFamilySamples()
+            );
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java:95]: nit: We could probably merge these into one line. 😉 
```suggestion
            TextFormat.writeFormat(format, writer, collectorRegistry.metricFamilySamples());
```

File: core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java [hunk 2]
```diff
@@ -111,4 +115,30 @@ void prometheusRequests() throws InterruptedException {
         verify(logger, times(6)).isDebugEnabled();
         verify(logger, times(4)).debug(anyString(), any(), any());
     }
+
+    @Test
+    void prometheusRequestsPrometheusFormat() throws InterruptedException {
+        final WebClient client = WebClient.of(server.httpUri());
+        final HttpRequest request = HttpRequest.builder()
+            .get("/enabled")
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java:123]: Could you fix the indentation? 😄 
You might refer to this document to set up the IDE.
https://armeria.dev/community/developer-guide#setting-up-your-ide

File: core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java [hunk 3]
```diff
@@ -111,4 +115,30 @@ void prometheusRequests() throws InterruptedException {
...
+        assertThat(
+            response.headers().get(HttpHeaderNames.CONTENT_TYPE)
+        ).isEqualTo(TextFormat.CONTENT_TYPE_004);
+    }
+
+    @Test
+    void prometheusRequestsOpenMetricsFormat() throws InterruptedException {
+        final WebClient client = WebClient.of(server.httpUri());
+        final HttpRequest request = HttpRequest.builder()
+            .get("/enabled")
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java:136]: ditto

**Review by minwoox — APPROVED** [review]
Left some minor comments.
Thanks a lot, @AngerM! 😄

File: core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java (hunk 2 above)
Comment (AngerM) [reply core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java:123]: hmm, importing those settings doesn't seem to be working for me in Intellij 2021.2. What's the format you are expecting? Is there a gradle command I can run to do autoformatting?

Comment (jrhee17) [reply core/src/test/java/com/linecorp/armeria/server/metric/PrometheusExpositionServiceTest.java]: If you've already the imported the settings, did you also try [Reformatting](https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html)?

File: core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java [hunk 4]
```diff
@@ -43,8 +44,6 @@
  */
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java:44]: Could you update the Javadoc? 😉
https://github.com/line/armeria/blob/198c247f2e9092601e920f900412f3cd1c70d7e0/core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java#L41-L42

**Comment by AngerM** [comment]
FYI I am away from my laptop for the next few weeks. Will take a look at the CR comments when I get back.

File: core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java [hunk 5]
```diff
@@ -87,18 +86,20 @@ public PrometheusExpositionService(CollectorRegistry collectorRegistry) {
 
     @Override
     protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        final String accept = req.headers().get(HttpHeaderNames.ACCEPT);
+        final String format = TextFormat.chooseContentType(accept);
         final ByteBuf buffer = ctx.alloc().buffer();
         boolean success = false;
         try (ByteBufOutputStream byteBufOutputStream = new ByteBufOutputStream(buffer);
              OutputStreamWriter writer = new OutputStreamWriter(byteBufOutputStream)) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java:94]:
```suggestion
            OutputStreamWriter writer = new OutputStreamWriter(byteBufOutputStream)) {
```

**Comment by minwoox** [comment]
Thanks a lot, @AngerM! 🎉 🎉 🎉

**Pull request merged**